### PR TITLE
Disk Space: Check based on actually required usage

### PR DESCRIPTION
**Pull request by tombettany**
Use `apt` to provide details about how much disk space is required to
complete the update. Provide an allowance for pip packages and buffer to
ensure that it doesn't completely fall apart if the numbers are slightly
off.

Needs some more testing but this is the base idea. What do you guys think?

### CODE REVIEW

**Comment by tombettany** [comment]
@radujipa Fixed the test here, all should work fine now.

Just the remaining problem that I was talking about this morning about how we deal with the postupdate scenario where we [install extra packages](https://github.com/KanoComputing/kano-updater/blob/master/kano_updater/scenarios.py#L781-L804). Moving forward we should have a better mechanism for apps in the app store but for now we can either add the packages that these guys depend on to the `kano-os` metapackage so that we know that they are sceduled for install at this point and so can account for them or remove this scenario entirely so that anybody new updating past 3.8.0 will no longer get them. Unless somebody else has a better idea.

**Comment by radujipa** [comment]
@tombettany Just a note on adding to the `kano-os` metapackage - I would avoid adding 3rd party apps to that package; I would be more in favour of grouping them in another, say, `kano-os-3rd-party` which would allow us to group any store apps we want there by default. Regardless, there was a point where the `kano-os` metapackage was installed in a postupdate IIRC so wouldn't moving them under it have the same issue?

File: kano_updater/disk_requirements.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+# disk_requirements.py
+#
+# Copyright (C) 2018 Kano Computing Ltd.
+# License: http://www.gnu.org/licenses/gpl-2.0.txt GNU GPL v2
+#
+# Variables pertaining to required disk usage for upgrade
+#
+
+
+PIP_REQ_SPACE = 100
```
Comment (radujipa) [inline kano_updater/disk_requirements.py]: Perhaps add a `# MB` comment to these?

File: kano_updater/apt_wrapper.py [hunk 2]
```diff
@@ -199,6 +203,21 @@ def packages_to_be_upgraded(self):
 
         return ret
 
+    def get_required_upgrade_space(self, priority=Priority.NONE):
```
[the remark this answers is not in the document]
Comment (tombettany) [reply kano_updater/apt_wrapper.py]: Good point here. In practice, we should only ever really end up with them either being marked for upgrade or keep but you're right, I'll add in a thing to make sure that the state is correctly reverted. Ideally I wouldn't have to mark them at all and could just look at the packages which are due for install but unfortunately the API doesn't let me. I'll take a look.

[the remark this answers is not in the document]
Comment (tombettany) [reply kano_updater/apt_wrapper.py]: I've addressed this by [storing the complete state at the start](https://github.com/KanoComputing/kano-updater/blob/2d8a6b23f3e728a74232548a774cf00391f3bbf4/kano_updater/apt_wrapper.py#L258-L269) adding a [mechanism for restoration](https://github.com/KanoComputing/kano-updater/blob/2d8a6b23f3e728a74232548a774cf00391f3bbf4/kano_updater/apt_wrapper.py#L23-L64).

**Comment by tombettany** [comment]
@radujipa I would definitely be happy with an additional `kano-os-3rd-party` metapackage. You are absolutely right about the `kano-os` metapackage being installed in the [update to 3.10.0](https://github.com/KanoComputing/kano-updater/blob/master/kano_updater/scenarios.py#L818-L823). Maybe the solution here should just be to add a clause that if the version is before 3.8.0 then require an additional 500MB or so (we bumped the disk space requirement by 500MB before for this update)

File: kano_updater/apt_wrapper.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ealdwulf) [reply kano_updater/apt_wrapper.py]: 👍

**Comment by tombettany** [comment]
@radujipa I've solved the problem via a [version check](https://github.com/KanoComputing/kano-updater/pull/278/commits/8246cee1a8a94bf8884a42ee26d29e81f273e909) which adds extra space requirement if upgrading from an older version.

**Comment by radujipa** [comment]
👍

**Pull request merged**